DataFormats/Portable/interface/PortableHostProduct.h

```diff
@@ -0,0 +1,99 @@
+#ifndef DataFormats_Portable_interface_PortableHostProduct_h
+#define DataFormats_Portable_interface_PortableHostProduct_h
+
+#include <cassert>
+#include <optional>
+
+#include <alpaka/alpaka.hpp>
+
+#include "HeterogeneousCore/AlpakaInterface/interface/config.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/host.h"
+#include "HeterogeneousCore/AlpakaInterface/interface/memory.h"
+
+// generic SoA-based product in host memory
+template <typename T>
+class PortableHostProduct {
+public:
+  using Product = T;
+  using Buffer = cms::alpakatools::host_buffer<Product>;
+  using ConstBuffer = cms::alpakatools::const_host_buffer<Product>;
+
+  PortableHostProduct() = default;
+
+  PortableHostProduct(alpaka_common::DevHost const& host)
+      // allocate pageable host memory
+      : buffer_{cms::alpakatools::make_host_buffer<Product>()}, product_{buffer_->data()} {
+    assert(reinterpret_cast<uintptr_t>(product_) % alignof(Product) == 0);
+  }
+
+  template <typename TQueue, typename = std::enable_if_t<alpaka::isQueue<TQueue>>>
+  PortableHostProduct(TQueue const& queue)
+      // allocate pinned host memory associated to the given work queue, accessible by the queue's device
+      : buffer_{cms::alpakatools::make_host_buffer<Product>(queue)}, product_{buffer_->data()} {
+    assert(reinterpret_cast<uintptr_t>(product_) % alignof(Product) == 0);
+  }
+
+  // non-copyable
+  PortableHostProduct(PortableHostProduct const&) = delete;
+  PortableHostProduct& operator=(PortableHostProduct const&) = delete;
+
+  // movable
+  PortableHostProduct(PortableHostProduct&&) = default;
+  PortableHostProduct& operator=(PortableHostProduct&&) = default;
+
+  // default destructor
+  ~PortableHostProduct() = default;
+
+  // access the product
+  Product& value() { return *product_; }
+  Product const& value() const { return *product_; }
+
+  Product* data() { return product_; }
+  Product const* data() const { return product_; }
+
+  Product& operator*() { return *product_; }
+  Product const& operator*() const { return *product_; }
+
+  Product* operator->() { return product_; }
+  Product const* operator->() const { return product_; }
+
+  // access the buffer
+  Buffer buffer() { return *buffer_; }
+  ConstBuffer buffer() const { return *buffer_; }
+  ConstBuffer const_buffer() const { return *buffer_; }
+
+  // part of the ROOT read streamer
+  static void ROOTReadStreamer(PortableHostProduct* newObj, Product& product) {
```

Comment:
Do I understand correctly this function is not currently called? Could you add a test where it is used?